### DISCUSSION TITLE
Smart blueprint: writing policies related to governance actions

### DISCUSSION THREAD

**Discussion by shlomitk1** [fybrik/fybrik#997] [Developers]
During the smart blueprint effort we remove the hard-coded logic existing in the manager, allowing instead to configure what capabilities to deploy, where and how to deploy them inside administrator policies written in Rego.
The topic of this discussion is how to configure deployment regarding governance actions, i.e. what modules will perform these actions, where these modules will be deployed, etc.
https://docs.google.com/document/d/1l7TmlacPjXGQJ_83OiJHYa40P_SL-YiQgxAcZqzyr_M/edit describes a way to configure that governance actions should run in geography of the dataset residency using the "transform" capability.
For example:
```
config[{"transform": decision}] {
    policy := {"ID": "transform-geo", "description":"Governance based transformations must take place in the geography where the data is stored"}
    clusters := [ data.clusters[i].name | data.clusters[i].metadata.region == input.request.dataset.geography ]
    decision := {"policy": policy, "restrictions": {"clusters": clusters}}
}
```
However, there are a number of issues with this approach.
1. Actions can be performed by other capabilities, such as read and copy.
2. By configuring a single capability, we don't have a global view on the constructed data path. 
For example: 
- a read-config module M1 runs in the workload cluster 
- a transform module M2 gets data via M1 api. It runs close to the data.
- a  read module M3 gets data from M2. It runs in the workload cluster.
To sum up, transformations were done close to the data, but M1 got access to the non-redacted data, which was not the intent.

Solution to 1. could be to separate read/write/copy from transform, when read/write/copy define interfaces, and transform define supported actions.
Perhaps, we should "flatten" action by making each action a capability. 

Another approach is to introduce a new abstraction for admin policies in addition to "config", for example: 
`access_raw_data = {"clusters": <restrictions on cluster region>}`
@simanadler @roee88 @roytman @ronenkat Please share your thoughts.

**Comment by simanadler** [comment]
@shlomitk1 In essence what we have today is a hierarchy of capabilities.  We have the high level capabilities (read, copy, write) and then those can have associated capabilities which we currently call actions.  Actions are specific types of data transforms.  This was built this way (I believe) because our current read and copy services can perform transforms as well.  

However, if one were to create a standalone module that does only transforms, then in today's architecture it's capability would be "transform" and the action would be the specific type of transform (redact, mask, ...).

The question is whether we should flatten the hierarchy and say transform is always a capability, e

> During the smart blueprint effort we remove the hard-coded logic existing in the manager, allowing instead to configure what capabilities to deploy, where and how to deploy them inside administrator policies written in Rego. The topic of this discussion is how to configure deployment regarding governance actions, i.e. what modules will perform these actions, where these modules will be deployed, etc. https://docs.google.com/document/d/1l7TmlacPjXGQJ_83OiJHYa40P_SL-YiQgxAcZqzyr_M/edit describes a way to configure that governance actions should run in geography of the dataset residency using the "transform" capability. For example:
> 
> ```
> config[{"transform": decision}] {
>     policy := {"ID": "transform-geo", "description":"Governance based transformations must take place in the geography where the data is stored"}
>     clusters := [ data.clusters[i].name | data.clusters[i].metadata.region == input.request.dataset.geography ]
>     decision := {"policy": policy, "restrictions": {"clusters": clusters}}
> }
> ```
> 
> However, there are a number of issues with this approach.
> 
>     1. Actions can be performed by other capabilities, such as read and copy.
> 
>     2. By configuring a single capability, we don't have a global view on the constructed data path.
>        For example:
> 
> 
>     * a read-config module M1 runs in the workload cluster
> 
>     * a transform module M2 gets data via M1 api. It runs close to the data.
> 
>     * a  read module M3 gets data from M2. It runs in the workload cluster.
>       To sum up, transformations were done close to the data, but M1 got access to the non-redacted data, which was not the intent.

So you are saying the order went read(workload cluster) -> transform (data cluster) -> read(workload cluster)
In essence we need to find a way to say in the policies that the transform **and capabilities it depends on** (i.e. run before it) need to be performed in the data cluster.
> 
> 
> Solution to 1. could be to separate read/write/copy from transform, when read/write/copy define interfaces, and transform define supported actions. Perhaps, we should "flatten" action by making each action a capability.

Yes, that was my thought.  In essence what we have today is a hierarchy of capabilities.  We have the high level capabilities (read, copy, write) and then those can have associated capabilities which we currently call actions.  Actions are specific types of data transforms.  This was built this way (I believe) because our current read and copy services can perform transforms as well and we don't have stand-alone transform modules currently.

However, if one were to create a standalone module that does only transforms, then in today's architecture it's capability would be "transform" and the action would be the specific type of transform (redact, mask, ...).

> 
> Another approach is to introduce a new abstraction for admin policies in addition to "config", for example: `access_raw_data = {"clusters": <restrictions on cluster region>}` @simanadler @roee88 @roytman @ronenkat Please share your thoughts.

Not sure I understand this option.

@roee88 What are your thoughts?
@zivnevo
